spirv_hlsl.cpp

```diff
@@ -0,0 +1,1061 @@
+	emit_op(result_type, result_id, join(op, "(", to_expression(op1), ", ", to_expression(op0), ")"), forward,
+	        false); // switch arguments instead of transposing
+
+	if (forward && forced_temporaries.find(result_id) == end(forced_temporaries))
+	{
+		inherit_expression_dependencies(result_id, op0);
+		inherit_expression_dependencies(result_id, op1);
+	}
+}
+
+void CompilerHLSL::emit_binary_func_op_transpose_all(uint32_t result_type, uint32_t result_id, uint32_t op0,
+                                                     uint32_t op1, const char *op)
+{
+	bool forward = should_forward(op0) && should_forward(op1);
+	emit_op(result_type, result_id, join(op, "(", to_expression(op1), ", ", to_expression(op0), ")"), forward,
+	        false); // switch arguments instead of double-transposing
+
+	if (forward && forced_temporaries.find(result_id) == end(forced_temporaries))
+	{
+		inherit_expression_dependencies(result_id, op0);
+		inherit_expression_dependencies(result_id, op1);
+	}
+}
+
+void CompilerHLSL::emit_uniform(const SPIRVariable &var)
+{
+	add_resource_name(var.self);
+	statement("uniform ", variable_decl(var), ";");
+}
+
+void CompilerHLSL::emit_glsl_op(uint32_t result_type, uint32_t id, uint32_t eop, const uint32_t *args, uint32_t count)
+{
+	GLSLstd450 op = static_cast<GLSLstd450>(eop);
+
+	switch (op)
+	{
+	case GLSLstd450InverseSqrt:
+	{
+		emit_unary_func_op(result_type, id, args[0], "rsqrt");
+		break;
+	}
+	case GLSLstd450Fract:
+	{
+		emit_unary_func_op(result_type, id, args[0], "frac");
+		break;
+	}
+	case GLSLstd450FMix:
+	case GLSLstd450IMix:
+	{
+		emit_trinary_func_op(result_type, id, args[0], args[1], args[2], "lerp");
+		break;
+	}
+	case GLSLstd450Atan2:
+	{
+		emit_binary_func_op(result_type, id, args[1], args[0], "atan2");
+		break;
+	}
+	default:
+		CompilerGLSL::emit_glsl_op(result_type, id, eop, args, count);
+		break;
+	}
+}
+
+void CompilerHLSL::emit_instruction(const Instruction &instruction)
+{
+	auto ops = stream(instruction);
+	auto opcode = static_cast<Op>(instruction.op);
+
+#define BOP(op) emit_binary_op(ops[0], ops[1], ops[2], ops[3], #op)
+#define BOP_CAST(op, type, skip_cast) emit_binary_op_cast(ops[0], ops[1], ops[2], ops[3], #op, type, skip_cast)
+#define UOP(op) emit_unary_op(ops[0], ops[1], ops[2], #op)
+#define QFOP(op) emit_quaternary_func_op(ops[0], ops[1], ops[2], ops[3], ops[4], ops[5], #op)
+#define TFOP(op) emit_trinary_func_op(ops[0], ops[1], ops[2], ops[3], ops[4], #op)
+#define BFOP(op) emit_binary_func_op(ops[0], ops[1], ops[2], ops[3], #op)
+#define BFOP_CAST(op, type, skip_cast) emit_binary_func_op_cast(ops[0], ops[1], ops[2], ops[3], #op, type, skip_cast)
+#define BFOP(op) emit_binary_func_op(ops[0], ops[1], ops[2], ops[3], #op)
+#define UFOP(op) emit_unary_func_op(ops[0], ops[1], ops[2], #op)
+
+	switch (opcode)
+	{
+	case OpMatrixTimesVector:
+	{
+		emit_binary_func_op_transpose_first(ops[0], ops[1], ops[2], ops[3], "mul");
```

Comment:
Separate helper function for this isn't needed now, can just use emit_binary_func_op directly with swapped arguments.